Semantics object support for edge triggered semantics

jonahwilliams opened the pull request:
Framework support for flutter/flutter#13704.  Also requires https://github.com/flutter/engine/pull/4901 in the engine first.


* Adds a `route: bool` parameter to Semantics, SemanticsProperties which corresponds to the SemanticsFlag.isRoute
 * Adds a route + explicitChildren semantic to cupertino and mv routes
 * Adds a route semantic to drawer
 * Adds a route semantic to the AlertDialog
  
Also updates a significant number of tests which asserted a specific semantics structure.

Commit:
initial commit

Commit:
add routeName semantics parameter

Commit:
remove debuggin info

Commit:
add basic doc comment to routeName

Commit:
fix doc comment in render_box

goderbauer commented on examples/flutter_gallery/lib/gallery/app.dart:
```diff
@@ -153,7 +153,7 @@ class GalleryAppState extends State<GalleryApp> {
       // https://docs.flutter.io/flutter/widgets/Navigator-class.html
       _kRoutes[item.routeName] = (BuildContext context) {
-        return item.buildRoute(context);
+        return new Semantics(routeName: item.routeName, child: item.buildRoute(context));
```
Will developers always have to customize their routes with Semantics widget to get proper announcements and audio-feedback for route changes? Can the framework detect (some/alll?) route changes automatically and produce the appropriate feedback to ensure that most route changes are accessibility-friendly out-of-the-box?

jonahwilliams replied:
To account for the differences between material/cupertino - this should probably be done at the widget level.  For example, in MaterialRoute I can add a semantics which defaults to the route name, or "home" for the initial route.  This can be overriden by the user by providing a more specific semantics deeper in the tree, but the announcement couldn't ever be removed unless we added a special sentinel value or something.

Another approach is to add a field to `RouteSettings`.  That might allow us to automatically handle the case of just using the route name, but still allow users to prevent announcements from being made.  This would also handle all of the different Route implementations at once.

For widgets like the drawer, It is probably best just to add another named argument with a reasonable default.

jonahwilliams replied:
we could also handle the lack of announcement names in iOS by pushing a sentinel string by default for iOS platforms, and then excluding that for cupertino widgets which announce themselves.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -147,6 +149,9 @@ class SemanticsData extends Diagnosticable {
   /// The reading direction is given by [textDirection].
   final String hint;
+  
+  /// The name of a widget subtree.
```
This and elsewhere need a stronger doc comment. What's the route name used for? What do TalkBack and VoiceOver do with it? When should this be set?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -147,6 +149,9 @@ class SemanticsData extends Diagnosticable {
   final String hint;
+  
+  /// The name of a widget subtree.
+  final String routeName;
```
Also: maybe we can shorten this to just "route"?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -91,6 +91,7 @@ class SemanticsData extends Diagnosticable {
     @required this.value,
     @required this.decreasedValue,
     @required this.hint,
+    @required this.routeName,
```
Should this also be made available to the CustomPainterSemantics? @yjbanov

yjbanov replied:
`CustomPaint` takes a `Widget child`, which can set it, or one of its ancestors can. `CustomPainter`s do not add any more widget subtrees otherwise. So my guess is, no? I can be convinced otherwise though.

goderbauer replied:
I don't have a strong opinion either and this can easily be added later. Just one more thought pro adding it (at some point):

Flutter is implemented using Custom Views on Android and iOS. It's possible that in the distant future somebody wants to implement a Flutter-like framework on top of Flutter (just like we do on top of Android/iOS) using Flutter's CustomPaint concept. They'll need a way to announce route names.

Hixie replied:
That's not a reasonable way to extend Flutter, I wouldn't worry about it. If you want to extend Flutter, you just do so right at the same layer we do.

goderbauer commented:
Can you please also add a test to ensure that the routeName given to a Semantics widget ends up on the appropriate semantics node?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -2612,6 +2647,7 @@ class SemanticsConfiguration {
       .._value = _value
```
I can't comment on the correct line, but `routeName` probably also needs to be handled in `absorb` and `isCompatibleWith` above.

Also, is it ok for the `routeName` to implicitly get merged into other semantics nodes? Or should a node that specifies a routeName always be a semantics boundary meaning that semantic information from that route cannot bubble up to parent nodes of that route?

Commit:
Merge remote-tracking branch 'upstream/master' into access

Commit:
update semantics to use route: bool param and fix all tests which depended on a specific semantic tree structure

Commit:
update documentation

Commit:
add examples to AlertDialogs

Commit:
Merge branch 'master' into access

jonahwilliams commented:
I've updated the PR with the new route behavior, as well as all of the tests which this broke

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +217,16 @@ class AlertDialog extends StatelessWidget {
+  ///
+  /// On android this would usually be a short summary of the contents. For
+  /// example - "App permissions" for a dialog containing permission
+  /// information.
```
Maybe add that this name is usually announced by TalkBack when the route is brought on screen.

jonahwilliams replied:
Done

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +217,16 @@ class AlertDialog extends StatelessWidget {
+  /// information.
+  ///
+  /// On iOS it is not common to provide a semantic name, since an announcement
+  /// chime is expected instead.
```
Maybe add that there are exception to this and list an example, e.g. Alerts?

jonahwilliams replied:
Rephrased this so that it is about dialogs and not the semantics in general

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -166,6 +166,7 @@ class AlertDialog extends StatelessWidget {
     this.content,
     this.contentPadding: const EdgeInsets.fromLTRB(24.0, 20.0, 24.0, 24.0),
     this.actions,
+    this.semanticName,
```
Not 100% what the native platforms do for an alert, but should this maybe default to "Alert"?

jonahwilliams replied:
done

goderbauer commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -100,13 +103,23 @@ class Drawer extends StatelessWidget {
+  /// See also:
+  ///
+  ///   * [Semantics.route]
+  final String semanticValue;
```
Is it intentional that it is called "semanticValue" here and "semanticName" in the AlertDialog? Could we unify these?

Also - and I know that naming is hard - could we find a more descriptive name for the property? Maybe something that tells the user that this is used as the route's name for semantics announcements?

jonahwilliams replied:
could also just go with routeName so that it is consistent with routes ... but I worry that will create confusion since this is not an actual route

goderbauer commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -237,7 +237,7 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
   @override
   Widget buildPage(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
-    final Widget result = builder(context);
+    final Widget result = new Semantics(route: true, explicitChildNodes: true, child: builder(context));
```
nit: formatting here and in many cases below: one line per property to make it easier to read?

jonahwilliams replied:
done

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +408,21 @@ class SemanticsProperties extends DiagnosticableTree {
+  /// Specificity is defined as the last child in a pre-order traversal - left
+  /// to right, root first.
+  ///
+  /// The announcement will contain the text in the [value] semantic.
```
Should this include that most platforms will still give an audible clue about a route change even if no value is provided?

jonahwilliams replied:
Added more details to docs

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +408,21 @@ class SemanticsProperties extends DiagnosticableTree {
+  /// Specificity is defined as the last child in a pre-order traversal - left
+  /// to right, root first.
+  ///
+  /// The announcement will contain the text in the [value] semantic.
```
Also I was wondering above: Is there a difference between setting value to `''` or `null`?

jonahwilliams replied:
No difference on android, neither gets announced.  Not sure on iOS.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +408,21 @@ class SemanticsProperties extends DiagnosticableTree {
   /// Doing so instructs screen readers to not read out the [value].
   final bool obscured;
 
+  /// If non-null, whether the node corresponds to a visually distinct subtree
```
full stop at the end?

jonahwilliams replied:
done

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -2336,6 +2352,17 @@ class SemanticsConfiguration {
     _hasBeenAnnotated = true;
   }
 
+  /// The name of a widget subtree.
```
Update this doc comment since route is a bool and not a name?

jonahwilliams replied:
done

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -306,11 +306,24 @@ class RouteSettings {
   ///
   /// If null, the route is anonymous.
+  ///
+  /// The is also used to create a default semantic value for the route
```
Does this work well since route names are usually hierarchically, e.g. "/home/login"? What would TalkBack or VoiceOver announce here?

jonahwilliams replied:
on Android I observed that TalkBalk splits these up - so "home/login" gets read as "home, login"  which seems like a reasonable default to me.

goderbauer replied:
I believe native Android would just announce "Login" in this case and not read the entire navigation stack? If we can't do better by default, we should at least give a pointer that tells you how to customize it if you want.

/cc @Hixie Do you have an opinion about reusing the (potentially structured) name as a default announcement?

jonahwilliams replied:
Some other options I explored - do string manipulation some home/login becomes [home, logic] and we only announce the last one.  That requires users to be using routes as we expect, and I don't really like magic string rules.

The other option is to add a new parameter to the route settings, but I'm not sure the best way to plumb that through, since usually we build route settings from a Map<String, WidgetBuilder>

jonahwilliams replied:
We could add a new param to the navigator called an "accessibilityBuilder" or something, which takes the routeSettings name and spits out the name used in the semantics node. This could default to splitting on"/" and returning the last segment

Hixie replied:
This is an internal opaque string. It doesn't make sense to use as a user-visible string.

goderbauer commented on packages/flutter/test/material/dialog_test.dart:
```diff
@@ -332,4 +334,47 @@ void main() {
+      ),
+    );
+
+    expect(semantics, isNot(includesNodeWith(flags: <SemanticsFlag>[SemanticsFlag.isRoute], value: 'alert')));
```
nit: I also find these slightly easier to read if the properties 'flags', and 'value' get their own line.

jonahwilliams replied:
done

goderbauer commented on packages/flutter/test/material/dialog_test.dart:
```diff
@@ -332,4 +334,47 @@ void main() {
     );
   });
+
+  testWidgets('Dialog widget contains route semantics', (WidgetTester tester) async {
```
Maybe add a test that ensures this when on iOS since this is one of the few cases where iOS also announces?

jonahwilliams replied:
I used a loop...hope that is okay for tests!

goderbauer commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -179,4 +183,12 @@ class MaterialPageRoute<T> extends PageRoute<T> {
   String get debugLabel => '${super.debugLabel}(${settings.name})';
+
+  String _routeName() {
+    if (_useCupertinoTransitions)
```
On iOS you can never provide a name to be announced for routes. Is that intentional?

jonahwilliams replied:
You can override this behavior by providing a child route semantics with a name.  The default is none since that reflects the behavior of accessibility in the iOS apps I have seen

Commit:
address some comments

Commit:
sync names and address more comments

Commit:
add more detail to semantics doc comment

goderbauer commented on examples/flutter_gallery/pubspec.yaml:
```diff
@@ -166,4 +166,10 @@ flutter:
       fonts:
         - asset: packages/flutter_gallery_assets/shrine/fonts/abrilfatface/AbrilFatface-Regular.ttf
 
+dependency_overrides:
```
You probably want to revert this :)

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +220,15 @@ class AlertDialog extends StatelessWidget {
+  /// Defaults to "alert" on both iOS and Android.
+  ///
+  /// This name is announced by TalkBack or VoiceOver when the dialog becomes
+  /// visible For example - "App permissions" for a dialog containing permission
```
Looks like this line is missing some kind of punctuation before "For example".

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -252,10 +265,14 @@ class AlertDialog extends StatelessWidget {
-          mainAxisSize: MainAxisSize.min,
-          crossAxisAlignment: CrossAxisAlignment.stretch,
-          children: children,
+        child: new Semantics(
```
Something that just crossed my mind while reading this: Can you make sure that this semantics node is not focusable with TalkBack on Android and VoiceOver? My suspicion is that you might actually be able to focus on that node because on both iOS and Android we say something is focusable when it has a value:
https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L877

https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L167

jonahwilliams replied:
You are right, this is focusing the page itself... I can fix this in the engine by excluding route flags from focusability.

jonahwilliams replied:
Fixed in https://github.com/flutter/engine/pull/4901/commits/6fcb204f80ab35c65429b399e8cc1cc4604eb382

goderbauer commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -100,13 +103,23 @@ class Drawer extends StatelessWidget {
       child: new Material(
         elevation: elevation,
-        child: child,
+        child: new Semantics(route: true, explicitChildNodes: true, value: value, child: child),
```
nit: multi-line?

jonahwilliams replied:
Done

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -306,11 +306,26 @@ class RouteSettings {
+  ///
+  /// See also:
+  ///
+  ///   * [Semantics.route]
```
This should say why I should look at Semantics.route.

jonahwilliams replied:
Done

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -306,11 +306,26 @@ class RouteSettings {
+  ///
+  /// See also:
+  ///
+  ///   * [Semantics.route]
```
same.

jonahwilliams replied:
Done

Commit:
fix some docs and remove committed pubspec

jonahwilliams commented:
Doesn't appear to be any difference between null and empty string - might have to do with how tonic encodes/decodes

Commit:
update routes to use accessibility name generator provided in navigator, and default implementation which splits on slashes

Commit:
remove unused member

jonahwilliams commented:
I've updated the navigator to take a parameter `typedef String RouteAccessibilityNameCallback(String routeName, bool isInitialRoute)`.  The default value handles splitting strings on "/", the default initial route name of "home", and iOS specific logic.  This information is computed when a route is pushed and placed into RouteSettings.  The material page then uses this info to construct a semantics node

Hixie commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -19,6 +19,9 @@ import 'theme.dart';
 // enum Department { treasury, state }
 
+/// The default for the [SemanticsProperty.value] of a dialog.
+const String _kDefaultAlertDialogValue = 'alert';
```
This has to be a localized string, we can't hard-code English.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -35,6 +35,24 @@ typedef bool RoutePredicate(Route<dynamic> route);
+      .lastWhere((String chunk) => chunk?.isNotEmpty == true, orElse: () => '');
+  }
+  return name;
+}
```
If we have this, it must be localized, not hard-coded to English.

We should definitely not expose the actual route names to users, those are internal details.

But I think we shouldn't have this at all. It's up to the developer to provide route names if they want them. We can't come up with a plausible default.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -288,6 +306,7 @@ class RouteSettings {
   /// Creates data used to construct routes.
   const RouteSettings({
     this.name,
+    this.accessibilityName,
```
The place where you create the settings won't ever know what the right name should be. This should be in the route's widget tree, not in the route settings.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -595,6 +628,20 @@ class Navigator extends StatefulWidget {
+  ///
+  ///   * [Semantics.route], the semantics flag which controls screen
+  ///     accessibility.
+  final RouteAccessibilityNameCallback generateAccessibility;
```
I don't understand how this API would be used. Why would this not just be part of the route's widget tree?

The place where you create the Navigator isn't where you would configure e.g. the app bar contents, so it shouldn't be where you configure the routes.

Hixie commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -83,6 +85,7 @@ class Drawer extends StatelessWidget {
   /// Typically used in the [Scaffold.drawer] property.
   const Drawer({
     Key key,
+    this.routeName = _kDefaultDrawerRouteNameSemantics,
```
we should call this something like "semanticLabel" or "semanticTitle" or just "title".
People using a drawer aren't going to really understand that it has anything to do with routes.

jonahwilliams replied:
I've removed this property from drawer - can be specified by having a child semantics node with a route/value.  For this and the other widgets, I've added default semantics which have a route flag and no name.

Maybe we can have some sensible defaults (maybe not) or provide a nicer way to give a semantic name to this route.  At any rate, it needs more discussion and doesn't have to happen now.

Hixie commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -100,13 +103,29 @@ class Drawer extends StatelessWidget {
   /// {@macro flutter.widgets.child}
   final Widget child;
 
+  /// An (optional) name for the drawer used by accessibility frameworks.
```
Not just accessibility frameworks, anything that uses semantics. For example, a search engine could inspect an app's contents using the semantics API. That's why it's not an accessibilty API, but a semantics API.

Hixie commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -100,13 +103,29 @@ class Drawer extends StatelessWidget {
+
   @override
   Widget build(BuildContext context) {
+    final String value = defaultTargetPlatform == TargetPlatform.iOS ? '' : routeName;
```
Do we really want to make this call (iOS vs Android) here?

General note: always use a switch when dealing with enums, not ?: or if statements. That way the analyzer will catch when we add a new platform and let us know everywhere that needs to change.

Hixie replied:
I guess it does make sense to make this call here.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -3217,6 +3219,16 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
   }
 
+  /// If non-null, sets the [SemanticsNode.route] semantic to the give value.
+  bool get route => _route;
```
"route", as a property name, sounds like it contains a route. This contains a bool, which doesn't match the name.

cc @goderbauer who ran into this problem with "password" recently and may have advice.

jonahwilliams replied:
I kept it this way for consistency with the other flags on `Semantics`, but I would have no problem renaming all of them to `isRoute`, `isPassword`, et cetera.  I could start with route in this CL and hit the other properties in a follow up, since that would be a breaking changed

jonahwilliams replied:
maybe `distinct`

Hixie replied:
I really think we have to bite the bullet and rename these. A "routeName" can't be a bool. That's just weird.

Hixie replied:
could be `namesRoute`?

Hixie replied:
and `scopesRoute` for the other?

Hixie replied:
cc @goderbauer

goderbauer replied:
Suggestions from @Hixie sound good to me.

jonahwilliams replied:
Done

Hixie commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +408,23 @@ class SemanticsProperties extends DiagnosticableTree {
+  ///
+  ///  * [SemanticsConfiguration.isRoute] for a description of how this is
+  ///    exposed in TalkBack and VoiceOver.
+  final bool route;
```
ditto

Hixie commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +408,23 @@ class SemanticsProperties extends DiagnosticableTree {
+  /// When the most specific route semantics node changes, the platform
+  /// accessibility framework will make an edge-triggered announcement and clear
+  /// the current accessibility focus. Specificity is defined as the last child
+  /// in a pre-order traversal - left to right, root first.
```
not sure what "left to right" means here, can you elaborate?

jonahwilliams replied:
Updated, right now it is the last child in inverse hit-test order

Hixie commented on packages/flutter/test/material/dialog_test.dart:
```diff
@@ -332,4 +335,58 @@ void main() {
     );
   });
+
+  for (TargetPlatform platform in <TargetPlatform>[TargetPlatform.iOS, TargetPlatform.android]) {
```
test all the platforms, not just iOS and Android

Hixie commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -122,6 +123,7 @@ class DialogDemoState extends State<DialogDemo> {
               showDemoDialog<DialogDemoAction>(
                 context: context,
                 child: new AlertDialog(
+                  routeName: 'location permissions',
```
why lowercase?

Hixie replied:
(as opposed to `Location permissions.`, say)

jonahwilliams replied:
Fixed

Hixie commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -237,7 +237,11 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
+      route: true,
+      explicitChildNodes: true,
+      child: builder(context)
+    );
```
same comments as for material page route.

jonahwilliams replied:
removed logic for semantics name in pages

Hixie commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +220,19 @@ class AlertDialog extends StatelessWidget {
+  /// See also:
+  ///   * [Semantics.route], the semantics flag which controls screen
+  ///     accessibility.
+  final String routeName;
```
same comment as elsewhere for "routeName"

jonahwilliams replied:
removed

Hixie commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -252,10 +269,14 @@ class AlertDialog extends StatelessWidget {
-          mainAxisSize: MainAxisSize.min,
-          crossAxisAlignment: CrossAxisAlignment.stretch,
-          children: children,
+        child: new Semantics(
```
seems like something we could push to `Dialog`, since all dialogs need this?

Hixie replied:
I guess AlertDialog is a special case because it has a "title". We should make sure we document in Dialog that you're expected to provide route name semantics.

Hixie replied:
maybe have a String argument to Dialog, say semanticTitle, that is normally null, and if non-null it sets the route name?
Then AlertDialog can do the right thing with just wrapping the title if it isn't given an explicit semanticTitle, and if it has a semanticTitle it can pass it on to Dialog and not worry about wrapping title.

Hixie commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -42,6 +42,8 @@ const double _kEdgeDragWidth = 20.0;
 const double _kMinFlingVelocity = 365.0;
 const Duration _kBaseSettleDuration = const Duration(milliseconds: 246);
 
+const String _kDefaultDrawerRouteNameSemantics = 'navigation menu';
```
can't be hardcoded

jonahwilliams replied:
removed

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -151,7 +151,12 @@ class MaterialPageRoute<T> extends PageRoute<T> {
+      explicitChildNodes: true,
+      value: settings.accessibilityName,
+      child: builder(context),
+    );
```
I could see putting the name here, though we'd need to put a constructor argument to configure it, and that would work poorly with things like the route table (MaterialApp.routes).

Other options we could consider (maybe do all of them):
 - We could have a `RouteName` widget, similar to `Title`, whose purpose is to populate the semantics tree with a route name.
 - We could have the AppBar automatically do this, e.g. label the contents of the appbar as "isRoute" (if I understand how this is configured on the engine side that would just work, right? cc @goderbauer).
- We could just tell people to use a Semantics widget like we do for dialogs.

jonahwilliams replied:
I think that at this point a minimal first pass seems to be the best thing to land.  I'll create a routeName widget and we can tell people to use semantics for the rest.  If there are issues, we can always add more defaults later - easier than trying to clean them up.

Labeling these as routes will work, a child semantics that is also labeled as a route with a value will get read by the engine implementations instead.

Hixie commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +220,19 @@ class AlertDialog extends StatelessWidget {
+  /// See also:
+  ///   * [Semantics.route], the semantics flag which controls screen
+  ///     accessibility.
+  final String routeName;
```
AlertDialog has a "title", which we really should just use for the route name. Nobody's ever going to specify the routeName.

Commit:
initial pass at addressing comments

Commit:
cleanup changes to unaffected tests

Commit:
add missing newline at eof

Commit:
remove mention of accessibility and ios default

Commit:
fix gallery examples

Commit:
update doc to specify inverse hit test order

Hixie commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -98,9 +98,14 @@ class DialogDemoState extends State<DialogDemo> {
+                  content: new Semantics(
+                    route: true,
+                    explicitChildNodes: true,
+                    value: 'Alert',
```
Won't the value automatically come from the child?

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,46 @@
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(new StringProperty('name', name, defaultValue: ''));
```
There's no default currently, so we should probably not specify one here.

But maybe we should have a default value?

The way I'd love to use this widget is just to wrap it with a single "child" argument, which automatically grabs the semantic value of the child and uses that as the default name. (Except maybe on iOS it should default to no value? I don't understand what iOS does exactly.)

jonahwilliams replied:
Spoke offline with @yjbanov  and @goderbauer  on this - it doesn't seem like that would give very good results as a default. For example, the child semantics are merged and contain multiple headers, or ordered in such a way that we can't determine what the right value to use is. 

And since we introducing a new Widget here and not inserting it into the navigator or pages, it makes sense to just require it.

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,46 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+/// A widget which provides a semantic name route name.
```
we should probably explain what this means. :-)

Hixie commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -123,9 +128,14 @@ class DialogDemoState extends State<DialogDemo> {
-                  content: new Text(
-                    _alertWithTitleText,
-                    style: dialogTextStyle
+                  content: new Semantics(
```
seems like a place to use RouteName

jonahwilliams replied:
reverted for now, label is inferred from title

Hixie commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -98,9 +98,14 @@ class DialogDemoState extends State<DialogDemo> {
-                  content: new Text(
-                    _alertWithoutTitleText,
-                    style: dialogTextStyle
+                  content: new Semantics(
```
can we use RouteName here?

jonahwilliams commented:
Another ... hiccup.  The platform specific logic for iOS changes between Dialogs (announced) and Pages/Drawers (chime).  In that case, it might make more sense to introduce `RouteName`, `DialogName` and `DrawerName` widgets for semantics information if this logic is to be handled at the framework/widget level.  But then, instead of doing all of that - why not just add a named arguments to all dialogs and drawers so that we can have a default semantic value.  RouteName could stay as a widget since routes aren't constructed in such a way where we have a clear root widget.

Which means that instead of pushing this off into another PR I should just rip off the bandaid now

Commit:
still in progress

Commit:
Merge branch 'master' of github.com:flutter/flutter into access

Commit:
Merge branch 'master' of github.com:flutter/flutter into access

Commit:
WIP

Commit:
WIP

jonahwilliams committed:
WIP

Commit:
initial work to add route semantics

Commit:
Merge branch 'master' of github.com:flutter/flutter into access

Commit:
merge

Commit:
WIP

Commit:
more cleanup, WIP

Commit:
Update project.pbxproj

Commit:
Update semantics.dart

Commit:
Update pubspec.yaml

Commit:
Update route_name.dart

Commit:
rename semantic name to label, remove extra member on popup

jonahwilliams commented:
@goderbauer , @Hixie  I've updated this PR to work with the engine PR.  Still have not updated tests - this will be a mostly mechanical but time consuming effort

Commit:
small cleanup

Commit:
fix custom painter and update custom painter and tooltip tests

Commit:
fix user accounts drawer header test

Hixie commented on packages/flutter/lib/src/material/drawer.dart:
```diff
@@ -100,13 +101,26 @@ class Drawer extends StatelessWidget {
+        child: new Semantics(
+          routeName: true,
+          route: true,
+          value: semanticLabel,
```
what happens when you set `routeName: true` but `value: null`?

Hixie commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +409,30 @@ class SemanticsProperties extends DiagnosticableTree {
+
+  /// If non-null, whether the node contains the semantic name for a route.
+  ///
+  /// When the most specific route semantics node changes, the platform will
```
"When the most specific route semantics node changes" could be interpreted many different ways. We should phrase this in a less ambiguous way.

Hixie commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +409,30 @@ class SemanticsProperties extends DiagnosticableTree {
+  /// If non-null, whether the node contains the semantic name for a route.
+  ///
+  /// When the most specific route semantics node changes, the platform will
+  /// search for the first child in inverse hit test order with a routeName 
```
hit test order is not fully defined, since it's geometric. Do you mean inverse paint order?

jonahwilliams replied:
Since I'm using the child list in the accessibility bridges, I believe the child lists have already been transformed?  I think yegor also has a pending change which updates the child order here.

The correct choice for scopes route is "what ever is on top", and the correct choice for names route is probably first in the expected semantic order.

https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L727

Hixie commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +409,30 @@ class SemanticsProperties extends DiagnosticableTree {
 
+  /// If non-null, whether the node corresponds to a visually distinct subtree.
+  ///
+  /// When the most specific route semantics node changes, the platform
```
"When the most specific route semantics node changes" is too ambiguous

Hixie commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4997,6 +5001,8 @@ class Semantics extends SingleChildRenderObjectWidget {
   /// This setting is often used in combination with [SemanticsConfiguration.isSemanticBoundary]
   /// to create semantic boundaries that are either writable or not for children.
+  /// 
+  /// Setting [properties.route] to true will force this value to be true as well.
```
This kind of magic is usually a bad sign... Can you talk more about this? Why is it necessary?

If we really need it then we should have a separate constructor for this case.

jonahwilliams replied:
Removed - I don't think this is necessary since most users will only interact with the route semantics through routeName.  Otherwise the correct way to handle this is probably in merging logic.

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,52 @@
+
+/// A widget which provides a semantic name for a route.
+///
+/// On iOS, no additional semantic information is inserted.
```
Talk more about why, and about what _does_ happen on Android.

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,52 @@
+
+  @override
+  Widget build(BuildContext context) {
+    if (defaultTargetPlatform == TargetPlatform.iOS) {
```
use a switch

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,52 @@
+    }
+    return new Semantics(
+      routeName: true,
+      explicitChildNodes: true,
```
isn't this implied?

Hixie commented on packages/flutter/test/widgets/route_name_test.dart:
```diff
@@ -0,0 +1,43 @@
+import 'dart:ui';
```
needs license

jonahwilliams replied:
Done

Hixie commented on packages/flutter/test/widgets/route_name_test.dart:
```diff
@@ -0,0 +1,43 @@
+
+import 'semantics_tester.dart';
+
+
```
extraneous blank line

Hixie commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,52 @@
+///  * [SemanticsProperties.routeName], for a description of how route name
+///    semantics work.
+class RouteName extends StatelessWidget  {
+
```
extraneous blank line

Hixie commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -237,7 +237,10 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
-    final Widget result = builder(context);
+    final Widget result = new Semantics(
+      route: true,
+      child: builder(context)
```
add trailling comma

Hixie commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -83,7 +92,13 @@ class Dialog extends StatelessWidget {
               color: _getColor(context),
               type: MaterialType.card,
-              child: child,
+              child: new Semantics(
```
here and elsewhere, why not use the RouteName widget?

jonahwilliams replied:
I didn't make RouteName insert a scopesRoute semantics, since most of the use will be providing names to what users think of as routes.  For Dialogs, Drawers, and PopupMenus we have a named parameter and default semantic values which I will add when I do the localizations - in addition to the inferred semantics from title like widgets

goderbauer commented on examples/flutter_gallery/lib/demo/pesto_demo.dart:
```diff
@@ -90,12 +90,14 @@ class _RecipeGridPageState extends State<RecipeGridPage> {
           },
         ),
-        body: new CustomScrollView(
+        body: new RouteName(
```
I would have expected the RouteName to wrap everything in the buildMethod?

goderbauer commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -382,6 +382,10 @@ class _AppBarState extends State<AppBar> {
     Widget title = widget.title;
     if (title != null) {
+      if (defaultTargetPlatform != TargetPlatform.iOS) {
+        title = new Semantics(routeName: true, child: title);
```
If the title widget is a string, it's value will end up as the label of a semantics node, not it's value (which according to the docs of isRouteName is used to announce a route)?

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -60,6 +61,14 @@ class Dialog extends StatelessWidget {
   /// Defaults to [Curves.fastOutSlowIn].
   final Curve insetAnimationCurve;
 
+  /// The semantic label of the dialog used by accessibility frameworks
```
Maybe be a little more descriptive right here how the value is used? E.g. to announce screen transitions when the dialog is brought on screen?

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -216,6 +232,14 @@ class AlertDialog extends StatelessWidget {
   /// from the [actions].
   final List<Widget> actions;
 
+  /// The semantic label of the dialog used by accessibility frameworks.
```
Same here.

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -443,6 +474,14 @@ class SimpleDialog extends StatelessWidget {
   /// the top padding ends up being 24 pixels.
   final EdgeInsetsGeometry contentPadding;
 
+  /// The semantic label of the dialog used by accessibility frameworks.
```
same.

goderbauer commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -606,6 +614,8 @@ class _PopupMenuRoute<T> extends PopupRoute<T> {
   @override
   final String barrierLabel;
 
+  final String semanticLabel;
```
This probably needs a doc comment.

goderbauer commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4997,6 +5001,8 @@ class Semantics extends SingleChildRenderObjectWidget {
   /// This setting is often used in combination with [SemanticsConfiguration.isSemanticBoundary]
   /// to create semantic boundaries that are either writable or not for children.
+  /// 
+  /// Setting [properties.route] to true will force this value to be true as well.
```
Where is this enforced? 

Should we maybe say something along the lines: If you set a `route` to true this also has to be true and then have an assert enforcing that Instead of magically setting it?

jonahwilliams commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -570,9 +615,9 @@ Future<T> showDialog<T>({
   WidgetBuilder builder,
 }) {
-  assert(child == null || builder == null); // ignore: deprecated_member_use
+  assert(child == null || builder == null);
```
Looks like the analyzer issue was fixed so they shouldn't be needed.

Commit:
address some comments

Commit:
initial pass renaming route and routename to scopesroute and namesroute

Commit:
more updates to docs and semantics

Commit:
update to gallery

Commit:
Update pubspec.yaml

Commit:
route name fixes

jonahwilliams commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -98,9 +98,14 @@ class DialogDemoState extends State<DialogDemo> {
-                  content: new Text(
-                    _alertWithoutTitleText,
-                    style: dialogTextStyle
+                  content: new Semantics(
```
changes to use label and routeName widget should fix this

jonahwilliams commented on examples/flutter_gallery/lib/demo/material/dialog_demo.dart:
```diff
@@ -98,9 +98,14 @@ class DialogDemoState extends State<DialogDemo> {
-                  content: new Text(
-                    _alertWithoutTitleText,
-                    style: dialogTextStyle
+                  content: new Semantics(
```
I'll need to add localizations first

jonahwilliams commented on examples/flutter_gallery/lib/gallery/app.dart:
```diff
@@ -153,7 +153,10 @@ class GalleryAppState extends State<GalleryApp> {
       _kRoutes[item.routeName] = (BuildContext context) {
-        return item.buildRoute(context);
+        return new RouteName(
+          routeName: item.routeName,
```
reverted

jonahwilliams commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -386,7 +386,10 @@ class _AppBarState extends State<AppBar> {
         softWrap: false,
         overflow: TextOverflow.ellipsis,
-        child: title,
+        child: new Semantics(
```
Fixed

jonahwilliams commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -70,7 +70,7 @@ class SemanticsTag {
 /// Summary information about a [SemanticsNode] object.
 ///
-/// A semantics node might [SemanticsNode.mergeAllDescendantsIntoThisNode],
+/// A semantics node might [SemanticsNode.fAllDescendantsIntoThisNode],
```
Fixed

Commit:
update routeName to introduce new route scope and update dialog, drawer, popup to use route name

Commit:
update and synchronize doc comments

Commit:
Update route_name_test.dart

Commit:
Merge branch 'master' of github.com:flutter/flutter into access

Commit:
Merge branch 'access' of github.com:jonahwilliams/flutter into access

Commit:
update all test semantics, move RouteName widget in drawer to avoid needing to check for null child

Commit:
Merge branch 'master' of github.com:flutter/flutter into access

Commit:
update default semantic label to not take precedence over title, remove extra assert, fix test

Commit:
undo pubspec change again

Commit:
fix import location

Commit:
cleanup to doc comments

Commit:
clean up tests and add test case for popup

jonahwilliams commented:
I've updated this PR with the default localized semantics labels, and cleaned up the documentation to reduce repetition - an explanation of scopesRoute and namesRoute now points to dart:ui.

goderbauer commented on examples/flutter_gallery/lib/demo/pesto_demo.dart:
```diff
@@ -78,25 +78,28 @@ class _RecipeGridPageState extends State<RecipeGridPage> {
-            _buildAppBar(context, statusBarHeight),
-            _buildBody(context, statusBarHeight),
-          ],
+    return new RouteName(
```
nit: the real fix here would probably be to give the Pesto logo on top of this page a label "pesto" and mark that one as routeName. Currently the title logo of the page is not focusable -- but that has nothing to do with this PR.

goderbauer commented on packages/flutter/lib/src/widgets/route_name.dart:
```diff
@@ -0,0 +1,67 @@
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+        result = new Semantics(
+          scopesRoute: true,
```
It also scopes a route? The doc above made it sound like it only gives a name to a surrounding route?

goderbauer replied:
Also: If I now use this within a MaterialRoute to give the route a custom name wouldn't there be two scopeRoutes in the tree for the same route?

goderbauer replied:
Having said this, if this doesn't specify scopesRoute, how would we ensure that this node doesn't become focusable?

jonahwilliams replied:
The "last" scopes route flag is selected, so this route name will take precedence.  This allows users to name custom routes and makes this useful for dialogs and drawers too

jonahwilliams replied:
Also update doc

goderbauer replied:
But wouldn't the semantic tree make it appear as if there are two nested routes on screen? I know it's fine for our current platforms, but what if other platforms decide to announce every route on screen?

jonahwilliams replied:
I think that either approach is okay for now, as it won't impact users building a standard cupertino/material style application.  But I also agree it would be best not to introduce redundant semantic information.

In the case of custom navigator/pages, having the RouteName also scope a route _might_ make it slightly easier to provide a working implementation.  But again, the implementation of namesRoute means that we don't have to.

Having them scope a route also let's me reuse the widget for Drawer, Dialog, and PopupMenu - but since we maintain these widgets I don't think it is a big deal for us to just directly add the correct semantic flags there.

@Hixie originally proposed using the route name for the other widgets, I'm not sure if he had a different idea from what I cahnged

jonahwilliams replied:
Oh right, the other issue is that the `RouteName` widget adds semantic information but doesn't really correspond to a focusable element, and it probably shouldn't just get merged with one.  And since we don't control focus on the framework side, I can  piggyback on the focus logic I added for scopesRoute to make sure it is ignored correctly

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -407,6 +409,25 @@ class SemanticsProperties extends DiagnosticableTree {
   final bool obscured;
 
+  /// If non-null, whether the node corresponds to the root of a subtree for
+  /// which values should be announced.
```
... for which a route name should be announced?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -2336,6 +2357,25 @@ class SemanticsConfiguration {
+  /// 
+  /// See also:
+  ///  * [SemanticsFlag.scopesRoute], for a full description of route scoping.
+  bool get scopesRoute => _hasFlag(SemanticsFlag.scopesRoute);
```
Maybe I missed it, but should there be an assert somewhere to ensure that you have explicitChildren set when you set this? Or are there use cases where it is actually ok to not do that?

jonahwilliams replied:
I've add an assert to the semantics render object, since that seems like the right place (can't use the const constructor)

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -60,12 +61,33 @@ class Dialog extends StatelessWidget {
+  /// announce screen transitions when the dialog is opened and closed.
+  /// 
+  /// If this label is not provided, it will default to
+  /// [MaterialLocalizations.dialogLabel].
```
Did you add the localizations in a separate PR? Or should they be part of this PR?

jonahwilliams replied:
I added them in a separate PR -https://github.com/flutter/flutter/pull/16633

Commit:
address comments

Commit:
update flexible space app bar and sliver app bar, refactor dialog semantics, simplify route name to only add namesRoute flag

Commit:
come on pubspec...

Commit:
remove extra label in pesto and remove duplicated route name in sliver app bar, fix test

Commit:
remove routeName widget

Commit:
remove gen files

Commit:
remove more gen files

goderbauer commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -3657,6 +3657,7 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
     super.describeSemanticsConfiguration(config);
     config.isSemanticBoundary = container;
     config.explicitChildNodes = explicitChildNodes;
+    assert((scopesRoute == true && explicitChildNodes == true) || scopesRoute != true);
```
The assert should probably include a message explaining that you have to set explicitChildNodes to true when scropesRoute is true.

jonahwilliams replied:
Done

goderbauer reviewed (approved):
LGTM

jonahwilliams committed:
add warning message to assert

jonahwilliams committed:
Merge remote-tracking branch 'upstream/master' into access

jonahwilliams committed:
Merge remote-tracking branch 'upstream/master' into access

jonahwilliams committed:
remove todos from engine roll

jonahwilliams committed:
remove recursive import of material

jonahwilliams committed:
fix import again and add missing new

jonahwilliams committed:
Merge remote-tracking branch 'upstream/master' into access
